20240927-multiple-object-deletion.md

```diff
@@ -0,0 +1,174 @@
+
+- Database overhead: Each deletion involves the execution of a separate SQL query, increasing the load on the database as the object count grows. Implementing bulk deletions gives us the opportunity to batch queries or delete multiple objects in a single query, reducing database overhead.
+
+For further context, the gateway currently supports Amazon S3's [DeleteObjects](https://docs.aws.amazon.com/AmazonS3/latest/API/API_DeleteObjects.html) operation, which allows for the deletion of multiple objects with a single HTTP request. However, the lack of native support for bulk deletions in the metainfo RPC service makes this operation subject to the aforementioned issues, as the gateway must translate every S3 request to a series of RPC requests.
+
+### Goals
+
+The goal of the implementation described by this document is to allow multiple objects to be deleted in a single request to the satellite.
+
+### Approach / Design
+
+#### Satellite
+
+The satellite metainfo RPC service must be extended with an endpoint allowing for the deletion of multiple objects in a single request.
+
+Requests directed to this endpoint must contain the following information:
+
+- The requester's API key
+- The bucket from which objects will be deleted
+- An indication of whether governance mode Object Lock restrictions should be bypassed
+- An indication of whether quiet mode should be enabled.
+  - With quiet mode enabled, only errors are returned; information about successful deletions is omitted from responses. Supporting this mode would allow us to save bandwidth on clients that use it.
+- A list of up to 1000 objects to delete
+
+The endpoint must respond with the following information:
+
+- A list containing the details of each successful deletion, including:
+  - The deleted object's key
+  - The deleted object's version (if the version wasn't specified in the request)
+  - The version ID of the delete marker (if applicable)
+- A list containing the details of each failed deletion, including:
+  - An error code
+  - The object's key
+  - The object's version ID (if it was provided in the request)
+
+Each bulk object deletion request will count as a single operation against the requester's deletion rate limit as opposed to each individual object deletion being counted separately.
+
+#### Libuplink
+
+Libuplink must provide a method through which the satellite metainfo RPC service's multiple object deletion endpoint is contacted. This is necessary for the gateway to function.
+
+#### Gateway
+
+The gateway must be given a configuration option that indicates whether the new multiple object deletion feature should be used. When it is enabled, it must invoke the new Libuplink method.
+
+#### Protobuf
```

Comment:
Any reason uplink cannot send a batch delete request to satellite instead?

Reply:
In my proposal, Uplink is still responsible for sending requests to the satellite; the Gateway <-> Uplink <-> Satellite request flow remains the same. Could you tell me what part of my document suggests otherwise so that I can make it clearer?

Reply:
I'm saying that the protocol already supports sending multiple deletes in a single request; so changes to protobuf shouldn't be necessary.

Reply:
Thank you for the explanation. In this case, while our protobuf schemas would remain unchanged, we would lose the ability to support quiet mode (and the bandwidth savings it brings) at the satellite level since it cannot be specified in individual deletion requests. Do you think this is worth it?

Reply:
If we need quiet, we could add that to the DeleteObject as well.

Reply:
But, yeah, one of the questions is that the behaviour might end up differing between S3 and our implementation... How does S3 handle failures -- does it abort the operation on the first error, or does it just delete as much it can?

Reply:
I'm pretty sure it tries everything and returns all the deletion errors, since each result in the response can have an error.

Reply:
Yeah, our batch handling aborts on the first error... although, it probably would be valid behaviour.

Reply:
To be clear, are you advising that I should minimize the amount of protobuf schema changes? This would require:
1. Extending `BeginDeleteObjectRequest` with a field indicating whether quiet mode is enabled, as you mentioned.
2. Extending `BeginDeleteObjectResponse` with fields carrying deletion failure information. For batch deletions, this would be used instead of responding with RPC errors for individual object deletion failures.

Reply:
If we can use the existing delete requests to do this, then that would be preferred. However, based on the discussion it's probably easier to make a new request&response

Reply:
Understood; thank you for your input. I have added your proposal to the "Alternatives considered" section.